pandas/src/hashtable_func_helper.pxi.in

```diff
@@ -112,3 +112,85 @@ def duplicated_{{dtype}}({{dtype}}_t[:] values,
     return out
 
 {{endfor}}
+
+#----------------------------------------------------------------------
+# Mode Computations
+#----------------------------------------------------------------------
+
+{{py:
+
+# dtype, ctype, table_type, npy_dtype
+dtypes = [('int64', 'int64_t', 'int64', 'int64'),
+          ('uint64', 'uint64_t', 'uint64', 'uint64'),
+          ('object', 'object', 'pymap', 'object_')]
+}}
+
+{{for dtype, ctype, table_type, npy_dtype in dtypes}}
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+
+{{if dtype == 'object'}}
+
+
+def mode_{{dtype}}(ndarray[{{ctype}}] values,
+                   ndarray[uint8_t, cast=True] mask):
+{{else}}
+
+
+def mode_{{dtype}}({{ctype}}[:] values):
+{{endif}}
+    cdef:
+        int count, max_count = 2
+        int j = -1 # so you can do +=
+        int k
+        kh_{{table_type}}_t *table
+        ndarray[{{ctype}}] modes
+
+    table = kh_init_{{table_type}}()
+
+    {{if dtype == 'object'}}
+    build_count_table_{{dtype}}(values, mask, table)
```

Comment:
`build_count_table_object` takes a `mask` argument, whereas the others do `null` detection internally and take a `dropna` parameter instead.  It might be possible to standardize, but as all of these functions are tied to Python-facing functions (e.g. `value_counts`), I don't think it would be a good idea to investigate in this PR.